sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyClient.java

```diff
@@ -1249,25 +1241,39 @@ private boolean isValidKeyLocallyAvailable() {
         }
 
         if (key == null && implClient.getKeyCollection() != null) {
-            key = CryptographyUtils.SECRETS_COLLECTION.equals(implClient.getKeyCollection())
-                ? getSecretKey() : getKey().getKey();
+            try {
+                // Get key from service and validate it. Then attempt to create a local cryptography client or default
+                // to service.
+                key = CryptographyUtils.SECRETS_COLLECTION.equals(implClient.getKeyCollection())
+                    ? implClient.getSecretKey() : getKey().getKey();
+            } catch (RuntimeException e) {
```

Comment:
I know exceptions in Java are not nearly as expensive as in .NET, but this is still using exceptions for control flow. Can you not instead handle this case to check for 403 or 404 like I do in .NET? It's an internal case, so perhaps you can use a non-throwing fetch.

Reply:
404 should be unlikely, but probably good to check. That the ID is the same in both cases is an implementation detail and not something we should rely on unconditionally.

Reply:
When you mean non-throwing fetch, do you mean to have the client that performs the get operation swallow the exception? I could change the internal code for getting the key for these crypto pperations to do that and have it return `null` or `Mono.empty()` instead, although that wouldn't change the fact that an exception would be generated by the client because we received a `403` or `404`.

On another note, if we got a `404` there's no way we can perform any operation (locally or on the service side), so I believe we should let things fail in that case and have the error message bubble up.

Reply:
This might be adumb question but is using exceptions for control flow a bad thing? I have seen it quite a few times in this and other libraries.

Reply:
Maybe it's idiomatic to Java, but using exceptions for control flow is taboo in other languages e.g., C++, C# (.NET), and others (though perhaps less taboo and more...why?).

Don't you get the actual response with the status code? In C# we do, and I just check if the status code is 403 (and maybe 404 - I forget) and, if so, I don't retrieve the private key. They get only the public key instead of the key pair. No exception is thrown anywhere in our stack. So I'm suggesting this client (both the sync and async clients) do that - not the caller.

Reply:
The Java SDKs are designed a bit differently where we have a centralized router in `azure-core` call `RestProxy` which consumes interfaces generated in the service SDKs (ex, `azure-security-keyvault-keys`) which uses the Swagger definition to determine which response statuses are considered "valid" or "good". In most cases a 404 will be invalid and this causes `RestProxy` to throw the defined exception type for that status code, so without jumping throw a lot of hoops there isn't a good way to handle this without an exception being thrown.

I will agree that using exceptions for control flow is bad, but this is a case where we don't have another option without a lot of complicated changes.

Reply:
Thanks for the explanation. At the very least, from a perf perspective, Java exceptions are ~4x less expensive than .NET's from our testing.